OneCommunityGlobal/HighestGoodNetworkApp
# Jerry - Track Who Changes Permissions On Permissions Management Page

# Description
<img width="696" alt="description" src="https://github.com/OneCommunityGlobal/HGNRest/assets/55968519/de743455-2395-4448-a35e-2d2d6289218f">

**Edit (Sep 21, 2023)**: Added a table on the Permissions Management page that fetches the permissionsChangeLogs from the dev database. Only Owner accounts can view see the table. Admin accounts are unable to see the table.


## Related PRS (if any):
* BE PR [#530](https://github.com/OneCommunityGlobal/HGNRest/pull/530)

## Main changes explained:

- Added `state.userProfile` to the request body of PATCH `/roles/:roleId` requests.

 **Edit (Sep 21, 2023)**:
- Table added to Permissions Management Page, only visible to Owner accounts.
- Table has pagination (20 rows per page), with pagination buttons. The page currently fetches ALL logs from permissionChangeLogs collection. If collection contains a really large number of logs, fetching logs may take a long time.


## How to test:

1. check into current branch. 
2. do `npm install` and `...` to run this PR locally
3. Go to BE PR and follow “How to test” instructions there to start the backend.
4. Log into a Owner account
5. From Dashboard, go to Other Links→ Permissions Management→ Jerry Test Role
    - Note: I advise you to test this branch only using “Jerry Test Role”, because it was created solely for testing purposes and will not interfere with the other roles needed for HGN app to run correctly. If you do insist on testing one of the other roles, I recommend testing the “Volunteer” role. 
    2. DO NOT test the “Owner” or “Admin" roles, as that will cause unwanted permissions changes to those vital roles.
6. Add/Delete permissions, then click Save.
7. In MongoDB dev database, check the `permissionChangeLogs` collection to see a log containing information about your latest permission change.
8. You may also check the `rolesMergedPermissions` collection to see that the changed role’s permissions field has also changed accordingly.

## Screenshots or videos of changes:

**Edit (Sep 21, 2023)**: 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/55968519/c587ba15-49d5-4292-9a4a-c2b5aa3ab0e6


**Edit (Sep 21, 2023)**: 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/55968519/0602b8e0-5837-4bf4-841e-d8b1d5b7e5a5




https://github.com/OneCommunityGlobal/HGNRest/assets/55968519/8fa9c22e-0fc3-46a4-a38c-955b9044201b


## Note:

It would be ideal for the `permissionChangeLogs` collection to be read-only and for the logs to have an expiration date (i.e., be automatically deleted after some set amount of time). From my understanding, these configurations can be made to the database by Jae or any trusted developer. If there it is possible/preferable to set these configurations in a Node.js environment, please let me know!
